### PR TITLE
[geometry] fix `LinearEntity.is_concurrent`

### PR DESCRIPTION
**Pull request by siefkenj**
The current implementation of `is_concurrent` fails if the first two lines don't intersect in a point.  However, the documentation states that `is_concurrent` returns true if and only if the `LinearEntity`s intersect in a single point.

```
 l1, l2 = Line((0,0),(1,1)), Line((0,0),(1,2))
 LinearEntity.is_concurrent(l1, l1, l1, l2)
```

should return `True`.

### CODE REVIEW

**Comment by smichr** [comment]
> should return True.

But l1 and l1 don't intersect in a single point so by definition, the answer should be False, I believe. The algorithm does need to be updated, however, since the first two lines in `Line.are_concurrent(l1, l2, l1)` intersect at a point and that point is (obviously) on the 3rd line so the routine returns True. So I think the logic should be:

```
if not any(i.is_parallel(j) for i,j in subsets(lines, 2)):
    pt = lines[0].intersect(lines[1])
    return all(pt in i for i in lines)
```

File: sympy/geometry/point.py [hunk 1]
```diff
@@ -106,7 +106,7 @@ def __new__(cls, *args, **kwargs):
     is_Point = True
 
     def __contains__(self, item):
-        return item == self
+        return item in self.args
```
Comment (smichr) [inline sympy/geometry/point.py]: Why did you change this?

File: sympy/geometry/point.py [hunk 2]
```diff
@@ -338,6 +338,11 @@ def evalf(self, prec=None, **options):
         where the coordinates are evaluated as floating point numbers to
         the precision indicated (default=15).
 
+        Parameters
```
Comment (smichr) [inline sympy/geometry/point.py]: +1

File: sympy/geometry/point.py [hunk 3]
```diff
@@ -360,7 +365,7 @@ def evalf(self, prec=None, **options):
     n = evalf
 
     def intersection(self, o):
-        """The intersection between this point and another point.
+        """The intersection between this point and another GeometryEntity.
```
Comment (smichr) [inline sympy/geometry/point.py]: +1

File: sympy/geometry/point.py (hunk 1 above)
Comment (siefkenj) [inline sympy/geometry/point.py]: So that a point acts more like a list, since it's enumerable.  This way

    for x in Point(1,2,3):
        print(x in Point(1,2,3))

will print `True`.  It doesn't seem particularly relevant to this pull request though :-/

**Comment by siefkenj** [comment]
Does concurrent mean "intersect at a single point" or "transversely intersect at a single point"?

File: sympy/geometry/point.py (hunk 1 above)
Comment (smichr) [inline sympy/geometry/point.py]: ok. I'm +.01-- since Point is iterable itmakessense. I'm not sure if Point should be iterable.

Comment (siefkenj) [inline sympy/geometry/point.py]: I think

    p = Point(1,2)
    x,y = p
    l = list(p)
    p2 = Point(x**2 for x in p)

are pretty good reasons for a point to be iterable.

**Comment by smichr** [comment]
> Does concurrent mean "intersect at a single point"

I defer to someone with greater geometrical fu. All I have seen for straight lines is the likes of [this](https://en.wikipedia.org/wiki/Bisection#Concurrencies_and_collinearities) which doesn't even have"transverse" on the page.

File: sympy/geometry/point.py (hunk 1 above)
Comment (smichr) [inline sympy/geometry/point.py]: +1

**Comment by siefkenj** [comment]
So I cannot find anything that directly addresses the point, but it seems like a **set** of lines is concurrent if it intersects in a single point.  In this case, the intersection is necessarily transverse.  It looks like no one defines concurrent for a **sequence** of lines.  In other words, a question of concurrency implies no duplicates.  I don't know what the right thing to do is...should the function throw an error if there are duplicate lines?  Or should the sequence be interpreted as a set, all duplicates removed, and then a concurrency result returned?

**Comment by smichr** [comment]
It would be in keeping with the docstring to treat the input as a set and thus remove duplicates. This will not, however, prevent a return value of False for the case when to lines are colinear as in `Line((0,0),(1,1))` and `Line((0,0),(2,2))`. And of course parallel lines will cause the result to be False, too. But if it were changed that way, it would have been nice to emphasize that the args are being treated as a  set (even though this isn't enforced the way it is in Intersection where, if you send a non-set, you get an error).

**Comment by siefkenj** [comment]
@smichr In other words, you think the original PR was correct where `is_concurrent(Line((0,0),(1,1)), Line((0,0),(2,2)))` fails but `is_concurrent(Line((0,0),(1,1)), Line((0,0),(2,2)), Line((0,0),(1,0)))` succeeds (since in the second case, the lines are silently interpreted as a set)?

**Comment by smichr** [comment]
```python
l1 = Line((0,0),(1,1))
l2 = Line((0,0),(2,2))
l3 = Line((0,0),(1,0))
>>> Line.are_concurrent(l1, l2)
False
>> Line.are_concurrent(l1, l2, l3)
False
>> Line.are_concurrent(l2, l2, l3)
True
```

I would say they are both false because they don't *all* intersect at a single point. In the second case I am not following your logic in saying that it would fail because it is being interpreted as a set. Did you mean that `Line.are_concurrent(l2, l2, l3)` would be True because it is being interpreted as a set? If so, yes -- that is in keeping with the docstring which says that you get "True : if the set of linear entities are concurrent". Since `l2` and `l3` intersect at a Point, the result is True.

Note, too, that it is **are**`_concurrent` not **is**`_concurrent`

**Comment by siefkenj** [comment]
I think the results should be

```
    l1 = Line((0,0),(1,1))
    l2 = Line((0,0),(2,2))
    l3 = Line((0,0),(1,0))
    >>> Line.are_concurrent(l1, l2)
    False
    >>> Line.are_concurrent(l1, l2, l3)
    True
    >>> Line.are_concurrent(l2, l2, l3)
    True
```

since `l1.equals(l2) == True`.  I will revert the pull request.  My mistake on the `is_concurrent` vs. `are_concurrent`.  I think the naming conventions for `geometry` need a bit more strategy...but that's besides the point :-).

**Comment by smichr** [comment]
> since l1.equals(l2) == True.

How would one affirm that two identical lines do not intersect at a single unique point? Lines can intersect at 0, 1 or infinite points. Only the case of 1 intersection (actually, for 3 more more lines) is said to be concurrent; the case of 2 non-parallel lines trivially intersects at a single point.

For the same reason, I think `Point(1,1).is_collinear(Point(1,1))` should be False since there is no unique line that can be drawn through two identical points.

**Comment by siefkenj** [comment]
[Mathworld](http://mathworld.wolfram.com/Concurrent.html) defines concurrent for two or more lines as having a common point of intersection.  Collinear should mean _there exists a line on which the points lie_, so a point should be collinear with itself.  For that matter, I think a set of objects is concurrent should mean _there exists a point of intersection_ and so a line would be concurrent with itself, but I don't see anyone who uses this definition.

[An Inductive Manual of The Straight Line and The Circle](https://books.google.com/books?id=Ci2lxhEwmvkC&pg=PA18&dq=concurrent+line&hl=en&sa=X&ved=0ahUKEwih6ovzm6_NAhVSwWMKHZEoBSg4KBDoAQg7MAU#v=onepage&q=concurrent%20line&f=false) is actually explicit with its definition.  It defines two or more lines to be concurrent if they aren't the same line and they intersect in a single point.  On the other hand, it defines collinear for three or more points, which makes it always a non-trivial property, but is unlike other conventions.

Using the [Hough Transformation](https://en.wikipedia.org/wiki/Hough_transform) in 2D, collinearity and concurrence are dual with the definitions _a set of lines is concurrent if there is a unique point of intersection_ and _a set of points is collinear if there is a unique line containing all points_, and in 3D, concurrence of planes and coplanarity are dual.  However, collinearity and concurrence are also dual in 2D with respect to the Hough transformation under the definition of _a set of lines is concurrent if there is a point of intersection_ and _a set of points is collinear if there is a line containing all points_.  To me, this is the most natural definition, but I won't advocate for it since everyone seems to agree that concurrence implies a unique point of intersection.

For what it's worth, Maple goes to the other extreme and defines `AreConcurrent` as _three lines are concurrent if they are coplanar and have a unique intersection point_ and `AreCollinear` as _three points are collinear if they lie on a single line_.  I don't have access to Maple to test what happens if I put the same argument in twice, but I don't advocate for their definitions...

**Comment by jksuom** [comment]
> a set of lines is concurrent should mean there exists a point of intersection.

I also think this is the most natural definition (i.e,. non-empty intersection). I would even like to extend this to 'a family of lines...'. (An indexed family may have repeated elements.)

File: sympy/geometry/point.py (hunk 1 above)
Comment (siefkenj) [inline sympy/geometry/point.py]: The thought just occurred to me.  Would

```
def __contains__(self, other):
    return other in self.args or other == self
```
be appropriate?

**Comment by smichr** [comment]
So are all in favor of committing this as is?

**Comment by smichr** [comment]
> but I won't advocate for it since everyone seems to agree that concurrence implies a unique point of intersection

Sorry, but I have to ask: are you in disagreement (with everyone) then with this PR? As I see it, you are returning True for concurrency when there is more than one Point of intersection, i.e., when that intersection is a Line. (The FiniteSet of length 1 might contain a Point or Line as the intersection.)

**Comment by siefkenj** [comment]
@smichr `Line().is_FiniteSet == False`, so this doesn't change the behavior from that stated in the docstring.  I'm in support of this patch.

**Comment by smichr** [comment]
For some reason I thought it gave `{Line(...)}`. OK, so committing as is.
